pkg/controller/direct/cloudbuild/workerpool_controller.go

```diff
@@ -0,0 +1,394 @@
+		return fmt.Errorf("update workerpool status %w", err)
+	}
+	status.CreateTime = ToDateDashTime(updated.GetCreateTime())
+	status.UpdateTime = ToDateDashTime(updated.GetUpdateTime())
+	return setStatus(u, status)
+}
+
+func (a *Adapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	return nil, nil
+}
+
+// Delete implements the Adapter interface.
+func (a *Adapter) Delete(ctx context.Context) (bool, error) {
+	if a.resourceID == "" {
+		return false, nil
+	}
+	req := &cloudbuildpb.DeleteWorkerPoolRequest{Name: a.fullyQualifiedName(), AllowMissing: true}
+	op, err := a.gcpClient.DeleteWorkerPool(ctx, req)
+	if err != nil {
+		// likely a server bug. worker_pool can be successfully deleted.
+		if !strings.Contains(err.Error(), "(line 12:3): missing \"value\" field") {
+			return false, fmt.Errorf("deleting cloudbuildworkerpool %s: %w", a.fullyQualifiedName(), err)
+		}
+	}
+	err = op.Wait(ctx)
+	if err != nil {
+		// likely a server bug. worker_pool can be successfully deleted.
+		if !strings.Contains(err.Error(), "(line 12:3): missing \"value\" field") {
+			return false, fmt.Errorf("waiting delete cloudbuildworkerpool %s: %w", a.fullyQualifiedName(), err)
+		}
+	}
+	return true, nil
+}
+
+func (a *Adapter) fullyQualifiedName() string {
+	return fmt.Sprintf("projects/%s/locations/%s/workerPools/%s", a.projectID, a.location, a.resourceID)
+}
+
+func (a *Adapter) getParent() string {
+	return fmt.Sprintf("projects/%s/locations/%s", a.projectID, a.location)
+}
+
+func setStatus(u *unstructured.Unstructured, typedStatus any) error {
+	status, err := runtime.DefaultUnstructuredConverter.ToUnstructured(typedStatus)
+	if err != nil {
+		return fmt.Errorf("error converting status to unstructured: %w", err)
+	}
+
+	old, _, _ := unstructured.NestedMap(u.Object, "status")
+	if old != nil {
+		status["conditions"] = old["conditions"]
+		status["observedGeneration"] = old["observedGeneration"]
+	}
+
+	u.Object["status"] = status
+
+	return nil
+}
+
+func ValueOf[T any](p *T) T {
+	var v T
+	if p != nil {
+		v = *p
+	}
+	return v
+}
+
+// IsNotFound returns true if the given error is an HTTP 404.
+func IsNotFound(err error) bool {
+	return HasHTTPCode(err, 404)
+}
+
+// HasHTTPCode returns true if the given error is an HTTP response with the given code.
+func HasHTTPCode(err error, code int) bool {
+	if err == nil {
+		return false
+	}
+	apiError := &apierror.APIError{}
+	if errors.As(err, &apiError) {
+		if apiError.HTTPCode() == code {
+			return true
+		}
+	} else {
+		klog.Warningf("unexpected error type %T", err)
+	}
+	return false
+}
+
+// LazyPtr returns a pointer to v, unless it is the empty value, in which case it returns nil.
+// It is essentially the inverse of ValueOf, though it is lossy
+// because we can't tell nil and empty apart without a pointer.
+func LazyPtr[T comparable](v T) *T {
+	var defaultValue T
+	if v == defaultValue {
+		return nil
+	}
+	return &v
+}
+
+func ToDateDashTime(ts *timestamppb.Timestamp) *string {
```

Comment:
Nit: Maybe `ToRFC3339Time` ?  IIRC I think this drops the nanos also, not sure if that's something we want to do...

Reply:
Ah I see where the name comes from!  Maybe `ToOpenAPIDateTime` would be better in that case?  I think the bigger question is whether we should preserve the full timestamp as provided to us (e.g. convert the timestamp to a string) ... i.e. does it matter that we drop precision (the nanos)